<a href="https://colab.research.google.com/github/Yesenia-AriasC/Introduccion-Pandas-Python/blob/main/3_Filtrado_con_loc_y_iloc.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Abramos el archivo csv con el que trabajaremos en esta ocasión (quedará como un dataframe).

In [1]:
import pandas as pd

In [2]:
df_books = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/PYTHON CIENCIA DE DATOS/Pandas/bestsellers with categories.csv',sep=',',header=0)

In [8]:
# df_books

# Slicing

Filtra por registros o filas.

In [7]:
df_books[0:4]

Name              Author  User Rating  \
0            10-Day Green Smoothie Cleanse            JJ Smith          4.7   
1                        11/22/63: A Novel        Stephen King          4.6   
2  12 Rules for Life: An Antidote to Chaos  Jordan B. Peterson          4.7   
3                   1984 (Signet Classics)       George Orwell          4.7   

   Reviews  Price  Year        Genre  
0    17350      8  2016  Non Fiction  
1     2052     22  2011      Fiction  
2    18979     15  2018  Non Fiction  
3    21424      6  2017      Fiction

# Filtro por columna

In [9]:
df_books[['Name','Author','Year']]

Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  Year  
0                    JJ Smith  2016  
1                Stephen King  2011  
2          Jordan B. Peterson  2018  
3               George Orwell  2017  
4    National Geographic Kids  2019  
..                        ...   ...  
545               Jeff Kinney  2019  
546               Jen Sincero  2016  
547               Jen Sincero  2017  
548               Jen Sincero  2018  
549               Jen Sincero  2019  

[550 rows x 3 columns]

# Filtros en pandas

df ['Columna'] == 'argumento'

df['Columna']  >  número

df['Columna'].str.startswith('J')

In [10]:
df_books['Price'] == '17350'

0      False
1      False
2      False
3      False
4      False
       ...  
545    False
546    False
547    False
548    False
549    False
Name: Price, Length: 550, dtype: bool

## df [ filtro ]

In [11]:
df_books[df_books['Reviews'] == 17350]

Name    Author  User Rating  Reviews  Price  Year  \
0  10-Day Green Smoothie Cleanse  JJ Smith          4.7    17350      8  2016   

         Genre  
0  Non Fiction

# por qué usar loc e iloc?
Permiten filtrar datos de manera mas específica.

Entonces, ¿por qué loc e iloc ?
En los casos que queremos filtrar también por columna. 

En este caso ponemos:

```df.loc[filas, columnas]```

```df.iloc[filas, columnas]```

La diferencia entre estos, es que en loc, en las columnas utilizamos una lista de las que queremos, por su nombre, en iloc debemos hacerlo por índices.


__Loc filtra segun un label mientras que iloc lo hace mediante indices.__


# .loc | [Filtra según un label](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html)

In [12]:
df_books.loc[:]
#muestra todos los datos del dataFrame

Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price  Year        Genre  
0                    JJ Smith          4.7    17350      8  2016  Non Fiction  
1                Stephen King          4.6     2052     22  2011      Fiction  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
545               Jeff Kinney          4.9     9413      8  2019      Fiction  
546               Jen Sincero          4.7    14331      8  2016  Non Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[550 rows x 7 columns]

## Rango de filas tomando en cuenta el start y el end

In [15]:
df_books.loc[0:4]

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

## Filtrando por filas y columnas | df.loc[0:4, ['Name', 'Author'] ] -> filas, columnas

In [16]:
df_books.loc[0:4, ['Name', 'Author']]
#filtra los datos de la fila que va de 0 a 4 y de las columnas Name y Author

Name                    Author
0                      10-Day Green Smoothie Cleanse                  JJ Smith
1                                  11/22/63: A Novel              Stephen King
2            12 Rules for Life: An Antidote to Chaos        Jordan B. Peterson
3                             1984 (Signet Classics)             George Orwell
4  5,000 Awesome Facts (About Everything!) (Natio...  National Geographic Kids

### Modificar valores de una columna específica del dataFrame

In [17]:
df_books.loc[:, ['Reviews']] * -1
#multiplica por -1 todos los valores de la columna Reviews

Reviews
0     -17350
1      -2052
2     -18979
3     -21424
4      -7665
..       ...
545    -9413
546   -14331
547   -14331
548   -14331
549   -14331

[550 rows x 1 columns]

### Filtrar datos que cumplan una condición determinada

In [18]:
df_books.loc[:, ['Author']] == 'JJ Smith' 
#muestra la columna Author con True en los valores que cumplen la condicion y False para los que no la cumplen

Author
0      True
1     False
2     False
3     False
4     False
..      ...
545   False
546   False
547   False
548   False
549   False

[550 rows x 1 columns]

# .iloc | [Filtra mediante índices.](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html)

In [19]:
df_books.iloc[:] 
#muestra todos los datos del dataframe

Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price  Year        Genre  
0                    JJ Smith          4.7    17350      8  2016  Non Fiction  
1                Stephen King          4.6     2052     22  2011      Fiction  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
545               Jeff Kinney          4.9     9413      8  2019      Fiction  
546               Jen Sincero          4.7    14331      8  2016  Non Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[550 rows x 7 columns]

## Filtrar datos según los índices de las filas y las columnas | df.iloc[0:4, 0:2] -> filas, columnas

In [20]:
df_books.iloc[:4, 0:2] 
#muestra los datos de las filas que van de 0 a 3 y las columnas con indices 0 y 1

Name              Author
0            10-Day Green Smoothie Cleanse            JJ Smith
1                        11/22/63: A Novel        Stephen King
2  12 Rules for Life: An Antidote to Chaos  Jordan B. Peterson
3                   1984 (Signet Classics)       George Orwell

### Buscar un dato específico | df.iloc[\#fila, \#columna]

In [21]:
df_books.iloc[1,3] 
#muestra el dato alojado en la fila 1 columna 3

2052

# DIFERENCIA IMPORTANTE ENTRE LOC E ILOC

Más adelante veremos esto a detalle, pero lo que haremos es añadir filas a nuestro dataframe, pegando df_books bajo él mismo.

De lo anteriormente descrito quedamos con esto:

In [22]:
df_books.append(df_books) # 1100 filas

Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price  Year        Genre  
0                    JJ Smith          4.7    17350      8  2016  Non Fiction  
1                Stephen King          4.6     2052     22  2011      Fiction  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
545               Jeff Kinney          4.9     9413      8  2019      Fiction  
546               Jen Sincero          4.7    14331      8  2016  Non Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[1100 rows x 7 columns]

LOC llama al indice o label, como pegamos el dataframe con su copia, aparece la fila consultada 2 veces:

In [24]:
#llama al indice
df_books.append(df_books).loc[540]

Name         Author  User Rating  Reviews  Price  Year    Genre
540  Wonder  R. J. Palacio          4.8    21625      9  2013  Fiction
540  Wonder  R. J. Palacio          4.8    21625      9  2013  Fiction

como el indice de las filas van hasta el 549, "llamar la fila 550" es imposible aunque claramente ésta existe:

In [ ]:
#Salta Error
df_books.append(df_books).loc[550]

Con ILOC si podemos llamar las filas por su Número real y no por su índice

In [28]:
#llama al número real de fila
df_books.append(df_books).iloc[550:555]

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction